### ISSUE TITLE
[Bug]: `subcomponents` meta property always invalid with TypeScript Strict mode for React

### ISSUE THREAD

**Issue by hudochenkov** [storybookjs/storybook#23170]
### Describe the bug

When React projects TypeScript config has Strict mode enabled `subcomponents` meta property always shows TypeScript error for every subcomponent property. When Strict mode is disabled no error is happening.

<img width="530" alt="Screenshot 2023-06-21 at 12 53 52" src="https://github.com/storybookjs/storybook/assets/654597/8d43d6d8-35a7-406d-9533-e8c0eaa4c09b">
<img width="563" alt="Screenshot 2023-06-21 at 12 54 51" src="https://github.com/storybookjs/storybook/assets/654597/ed159f7e-6597-48e2-8bda-54f2801ef29c">

```
Type '({ children }: RequiredChildrenProps) => Element' is not assignable to type 'ComponentType<unknown>'.
  Type '({ children }: RequiredChildrenProps) => Element' is not assignable to type 'FunctionComponent<unknown>'.
    Types of parameters '__0' and 'props' are incompatible.
      Type 'unknown' is not assignable to type 'RequiredChildrenProps'.(2322)
```

### To Reproduce

https://stackblitz.com/edit/github-w3gat5?file=tsconfig.json,src%2Fstories%2FMainComponent.stories.tsx&preset=node

It is configured with `strict: true`.

### System

```shell
Environment Info:

  System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
  Binaries:
    Node: 16.14.2 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 9.4.2 - /usr/local/bin/npm
  npmPackages:
    @storybook/addon-essentials: ^7.1.0-alpha.37 => 7.1.0-alpha.37 
    @storybook/addon-interactions: ^7.1.0-alpha.37 => 7.1.0-alpha.37 
    @storybook/addon-links: ^7.1.0-alpha.37 => 7.1.0-alpha.37 
    @storybook/blocks: ^7.1.0-alpha.37 => 7.1.0-alpha.37 
    @storybook/react: ^7.1.0-alpha.37 => 7.1.0-alpha.37 
    @storybook/react-webpack5: ^7.1.0-alpha.37 => 7.1.0-alpha.37 
    @storybook/testing-library: ^0.1.1-next.0 => 0.1.1-next.0
```

### Additional context

Issue is not present in Storybook 6.

`subscomponents` type in Storybook 6:

```ts
subcomponents?: Record<string, React.ComponentType<any>> | undefined
```

`subscomponents` type in Storybook 7:

```ts
ComponentAnnotations<ReactRenderer, MainComponentProps>.subcomponents?: Record<string, React.ComponentType<unknown>> | undefined
```

Probably the main issue is a change from `React.ComponentType<any>` to `React.ComponentType<unknown>`.

**Comment by valentinpalkovic** [comment]
fyi: Please keep in mind that we might drop support for `subcomponents` altogether. Follow the discussion here: https://github.com/storybookjs/storybook/issues/20782

**Comment by hudochenkov** [comment]
Yes, I've stumbled upon that discussion. Since the `ArgsTable` and `subcomponents` is the only convenient solution for documenting compound components (e. g. `List` and `List.Item`), we'll have to use it.

**Comment by robindykema** [comment]
Hi, I'm still seeing this bug now that `subcomponents` support was added back in Storybook 8. I'm on version 8.0.6.

**Comment by jamilalisgandarov** [comment]
Still happening on 8.0.8

**Comment by erwinnurhaq** [comment]
same here. any news?

**Comment by fullofcaffeine** [comment]
+1

**Comment by fernandoabel** [comment]
As a workaround for the typescript error, I have done this in my project.
Hope that helps in the meantime.

`**import React, { ComponentType } from 'react';**
import type { Meta, StoryObj } from '@storybook/react';
import { MainComponent } from './MainComponent';

const meta: Meta<typeof MainComponent> = {
  title: 'Example/MainComponent',
  component: MainComponent,
  subcomponents: {
    **'MainComponent.RequiredChildren': MainComponent.RequiredChildren as ComponentType<any>,
    'MainComponent.OptionalChildren': MainComponent.OptionalChildren as ComponentType<any>,
    'MainComponent.PropAndChildren': MainComponent.PropAndChildren as ComponentType<any>,
    'MainComponent.OnlyProp': MainComponent.OnlyProp as ComponentType<any>,**
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const DefaultStory: Story = {};
`

**Comment by ecsbeats** [comment]
Still happening on v8.2.4

**Comment by lobaak** [comment]
This seems to work as a temporary solution until it's fixed

```typescript
subcomponents: {
  Component: Component as React.ComponentType<any>,
},
```

**Comment by oncet** [comment]
Using `v8.2.6` and the issue persists. Subcomponents was my main motivation for upgrading to v8.

**Comment by Abraham-Flutterwave** [comment]
Still happening

**Comment by penx** [comment]
> fyi/related: Please keep in mind that we might drop support for `subcomponents` altogether. Follow the discussion here: [#20782](https://github.com/storybookjs/storybook/issues/20782)

for anyone tracking this, the subcomponents feature has been undeprecated as detailed by @JReinhold here:

https://github.com/storybookjs/storybook/issues/20782#issuecomment-1892178858

Given this - any chance of taking a look at this issue or the [linked PR](https://github.com/ComponentDriven/csf/pull/103) @JReinhold @valentinpalkovic?

**Comment by JReinhold** [comment]
I believe this was fixed in https://github.com/ComponentDriven/csf/pull/103

**Comment by mihkeleidast** [comment]
@JReinhold this seems to have broken again with Storybook 9 - the type now uses internal Storybook core definition, not the csf package one. I've filed a PR to bring the fix into core as well, but pinging just in case you want to re-open this meanwhile.
